### ISSUE TITLE
Assert firing sporadically in ManagedHandler when returning connection to pool

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#26754]
https://github.com/dotnet/corefx/blob/95bfb895707d5756d4b9693286672780f0a6ba02/src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs#L1216
(I added the current values of the variables to the assert message.)
```
  FailFast: 34 != 32

     at System.Diagnostics.Debug.Assert(Boolean condition, String message, String detailMessage) in E:\A\_work\2298\s\src\mscorlib\shared\System\Diagnostics\Debug.cs:line 97
     at System.Net.Http.HttpConnection.ReturnConnectionToPoolCore() in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\Managed\HttpConnection.cs:line 1202
     at System.Net.Http.HttpConnection.ReturnConnectionToPool() in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\Managed\HttpConnection.cs:line 1164
     at System.Net.Http.HttpConnection.ContentLengthReadStream.ReadAsync(Memory`1 destination, CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\Managed\
  ContentLengthReadStream.cs:line 60
     at System.Runtime.CompilerServices.AsyncMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine) in E:\A\_work\2298\s\src\mscorlib\shared\System\Runtime\CompilerServices\AsyncMethodBuilde
  r.cs:line 38
     at System.Net.Http.HttpConnection.ContentLengthReadStream.ReadAsync(Memory`1 destination, CancellationToken cancellationToken)
     at System.Net.Http.HttpConnection.ContentLengthReadStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\Syst
  em\Net\Http\Managed\ContentLengthReadStream.cs:line 28
     at System.IO.Stream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count) in E:\A\_work\2298\s\src\mscorlib\src\System\IO\Stream.cs:line 376
     at System.Net.Http.Functional.Tests.ResponseStreamTest.ReadAsStreamHelper(IPEndPoint serverEndPoint) in d:\repos\corefx\src\System.Net.Http\tests\FunctionalTests\ResponseStreamTest.cs:line 253
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in E:\A\_work\2298\s\src\mscorlib\shared\System\Threading\ExecutionCon
  text.cs:line 166
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext() in E:\A\_work\2298\s\src\mscorlib\src\System\Runtime\CompilerServices\AsyncMethodBuilder.cs:line 546
     at Xunit.Sdk.AsyncTestSyncContext.<>c__DisplayClass7_0.<Post>b__1(Object _) in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\AsyncTestSyncContext.cs:line 75
     at Xunit.Sdk.MaxConcurrencySyncContext.RunOnSyncContext(SendOrPostCallback callback, Object state) in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\MaxConcurrencySyncContext.cs:lin
  e 107
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in E:\A\_work\2298\s\src\mscorlib\shared\System\Threading\ExecutionCon
  text.cs:line 166
     at Xunit.Sdk.MaxConcurrencySyncContext.WorkerThreadProc() in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\MaxConcurrencySyncContext.cs:line 89
     at Xunit.Sdk.XunitWorkerThread.<>c.<QueueUserWorkItem>b__5_0(Object _) in C:\BuildAgent\work\cb37e9acf085d108\src\common\XunitWorkerThread.cs:line 37
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in E:\A\_work\2298\s\src\mscorlib\shared\System\Threading\ExecutionCon
  text.cs:line 166
     at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot) in E:\A\_work\2298\s\src\mscorlib\src\System\Threading\Tasks\Task.cs:line 2440
     at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in E:\A\_work\2298\s\src\mscorlib\shared\System\Threading\ExecutionCon
  text.cs:line 166

     at System.Environment.FailFast(System.String, System.Exception)
     at System.Net.Http.HttpConnection.ReturnConnectionToPoolCore()
     at System.Net.Http.HttpConnection.ReturnConnectionToPool()
     at System.Net.Http.HttpConnection+ContentLengthReadStream+<ReadAsync>d__3.MoveNext()
     at System.Runtime.CompilerServices.AsyncMethodBuilder.Start[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.__Canon ByRef)
     at System.Net.Http.HttpConnection+ContentLengthReadStream.ReadAsync(System.Memory`1<Byte>, System.Threading.CancellationToken)
     at System.Net.Http.HttpConnection+ContentLengthReadStream.ReadAsync(Byte[], Int32, Int32, System.Threading.CancellationToken)
     at System.IO.Stream.ReadAsync(Byte[], Int32, Int32)
     at System.Net.Http.Functional.Tests.ResponseStreamTest+<ReadAsStreamHelper>d__9.MoveNext()
     at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
     at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToke
  n=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
     at Xunit.Sdk.AsyncTestSyncContext+<>c__DisplayClass7_0.<Post>b__1(System.Object)
     at Xunit.Sdk.MaxConcurrencySyncContext.RunOnSyncContext(System.Threading.SendOrPostCallback, System.Object)
     at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
     at Xunit.Sdk.MaxConcurrencySyncContext.WorkerThreadProc()
     at Xunit.Sdk.XunitWorkerThread+<>c.<QueueUserWorkItem>b__5_0(System.Object)
     at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
     at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
     at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
```

**Comment by benaadams** [comment]
```
   at System.Runtime.CompilerServices.AsyncMethodBuilder.Start
```
In that trace... its like trollin' me

**Comment by geoffkizer** [comment]
When are you seeing this?  Local runs?  CI?

**Comment by stephentoub** [comment]
Local

**Comment by stephentoub** [comment]
Yeah, repros locally with a clean master build, maybe once out of 10 outerloop runs.

**Comment by wfurt** [comment]
I'll take a look. Hopefully I'll be able to reproduce as well.

**Comment by pjanotti** [comment]
In my local hits for this it was always coming from `System.Net.Http.Functional.Tests.ManagedHandler_ResponseStreamTest.ReadAsStreamAsync_ValidServerResponse_Success(transferType: ContentLength, transferError: None)`. 

In the test code we are adding "\r\n" to the content, but those two are not part of the content length (which has length 30). If I am reading this right whenever `ContentLengthReadStream` assumes that it read all the content and tries to return the connection to the pool we hit the assert. 

@stephentoub @geoffkizer isn't the case that the expectation in the assert can't be guaranteed against real servers?

Anyway, in the short run I will modify the test so content-length matches the data on the stream to avoid the test instability. I will do a bunch of local runs before starting the PR to confirm if my theory holds.

**Comment by stephentoub** [comment]
Thanks for digging into it, @pjanotti.

Yeah, seems like there are two issues here:
1. The test is wrong. https://github.com/dotnet/corefx/blob/25c4a916bf1f35be1d009b679b19f7ebea535647/src/Common/tests/System/Net/Http/LoopbackServer.cs#L259 shouldn't include \r\n.
2. The assert is highlighting a real issue in the code.  As @pjanotti states, we can't trust the server will be honest in this regard, which means that we actually need to throw away any remaining data in the buffer, and potentially fail the request if there's more data in the buffer than was expected by Content-Length.  We should be able to have this assert here; if it were valid for data to remain in the buffer, the next request that uses this connection could end up using leftover data from the old one.  There's also a perf benefit to resetting the offset to 0 so that we have use of the entire buffer.

**Comment by geoffkizer** [comment]
Yep, agreed on all points.  We should also add tests that explicitly detect this case (server sends extra stuff, we fail the connection).

**Comment by stephentoub** [comment]
> We should also add tests that explicitly detect this case (server sends extra stuff, we fail the connection).

Yeah, although I think it will be difficult to test well.  There's a race condition here as to whether we receive the additional data as part of the same read and thus can tell prior to finishing the response stream and before returning the connection to the pool whether there's more data.  Otherwise it just ends up looking later like a connection we can't reuse because stray data was received.

**Comment by geoffkizer** [comment]
Yeah, but either way the connection shouldn't be reused.  I think we can add a test where we send a response with added junk, and then try to send another request and ensure that it happens on a second connection and the original connection is closed.

**Comment by stephentoub** [comment]
Yup.  It'll just be a small variation on the tests here: https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/tests/FunctionalTests/ManagedHandlerTest.cs#L318
